File: Search/python/13_유기농배추.py
# fast campus 강의
# https://www.acmicpc.net/problem/1012
# 1
import sys
from collections import deque
sys.setrecursionlimit(10000)

T = int(input())
matrix = [[]]
visit = [[]]
dx, dy = [0, 1, 0, -1], [1, 0, -1, 0]

def dfs(x, y):
    global matrix, visit
    visit[x][y] = True
    for i in range(4):
        mx, my = x + dx[i], y + dy[i]
        if matrix[mx][my] == 0 or visit[mx][my]:
            continue
        dfs(mx, my)

def bfs(x, y):
    q = deque([x, y])
    while q:
        px = q.popleft()
        py = q.popleft()
        if not visit[px][py]:
            visit[px][py] = True
            for i in range(4):
                mx = px + dx[i]
                my = py + dy[i]
                if matrix[mx][my] == 0 or visit[mx][my]:
                    continue
                q.append(mx)
                q.append(my)

def process():
    global matrix, visit
    M, N, K = map(int, input().split())
    matrix = [[0 for i in range(M + 2)] for _ in range(N + 2)]
    visit = [[False for i in range(M + 2)] for _ in range(N + 2)]
    for _ in range(K):
        x, y = map(int, input().split())
        matrix[y + 1][x + 1] = 1
    
    result = 0
    for i in range(1, N + 1):
        for j in range(1, M + 1):
            if matrix[i][j] == 0 or visit[i][j]:
                continue
            bfs(i, j)
            result += 1
    print(result)


for _ in range(T):
    process()

File: Search/c++/13_유기농배추.cpp
// fast campus 강의
// https://www.acmicpc.net/problem/1012
// 2

#include <iostream>
#include <queue>
#include <cstring>

using namespace std;

int T, N, M, K;
int matrix[51][51];
bool visited[51][51];
int mx[4] = {1, 0, -1, 0};
int my[4] = {0, 1, 0, -1};

void dfs(int x, int y) {
    visited[x][y] = true;
    for (int i = 0; i < 4; i++) {
        int nx = x + mx[i];
        int ny = y + my[i];
        
        if (nx < 0 || nx >= N || ny < 0 || ny >= M)
            continue;
        if (!visited[nx][ny] && matrix[nx][ny])
            dfs(nx, ny);
    }
}

void bfs(int x, int y) {
    queue<pair<int, int>> q;
    q.push({x, y});
    visited[x][y] = true;

    while (!q.empty()) {
        int px = q.front().first;
        int py = q.front().second;
        q.pop();

        for (int i = 0; i < 4; i++) {
            int nx = px + mx[i];
            int ny = py + my[i];

            if (nx < 0 || nx >= N || ny < 0 || ny >= M)
                continue;
            
            if (!visited[nx][ny] && matrix[nx][ny]) {
                q.push({nx, ny});
                visited[nx][ny] = true;
            }
        }
    }
}

int main() {
    cin >> T;
    for (int t = 0; t < T; t++) {
        memset(matrix, 0, sizeof(matrix));
        memset(visited, false, sizeof(visited));

        cin >> N >> M >> K;
        for (int i = 0; i < K; i++) {
            int x, y;
            cin >> x >> y;
            matrix[x][y] = 1;
        }

        int result = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (!visited[i][j] && matrix[i][j]) {
                    result += 1;
                    dfs(i, j);
                }
            }
        }
        cout << result << '\n';
    }
}